dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -54,34 +56,50 @@
     public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
         SettableFuture<Response> future = SettableFuture.create();
 
-        Supplier<ListenableFuture<Response>> callSupplier = () -> delegate.execute(endpoint, request);
-        FutureCallback<Response> retryer = new RetryingCallback<>(callSupplier, future);
-        Futures.addCallback(callSupplier.get(), retryer, DIRECT_EXECUTOR);
+        Function<Integer, ListenableFuture<Response>> callSupplier = attempt -> {
+            // TODO(dfox): include retry number in the request somehow
+            return delegate.execute(endpoint, request);
+        };
+        FutureCallback<Response> retryer = new RetryingCallback(callSupplier, future);
+        Futures.addCallback(callSupplier.apply(0), retryer, DIRECT_EXECUTOR);
 
         return future;
     }
 
-    private final class RetryingCallback<T> implements FutureCallback<T> {
+    private final class RetryingCallback implements FutureCallback<Response> {
         private final AtomicInteger failures = new AtomicInteger(0);
-        private final Supplier<ListenableFuture<T>> runnable;
-        private final SettableFuture<T> delegate;
+        private final Function<Integer, ListenableFuture<Response>> runnable;
+        private final SettableFuture<Response> delegate;
 
-        private RetryingCallback(Supplier<ListenableFuture<T>> runnable, SettableFuture<T> delegate) {
+        private RetryingCallback(
+                Function<Integer, ListenableFuture<Response>> runnable, SettableFuture<Response> delegate) {
             this.runnable = runnable;
             this.delegate = delegate;
         }
 
         @Override
-        public void onSuccess(T result) {
+        public void onSuccess(Response result) {
+            // this condition should really match the BlacklistingChannel so that we don't hit the same host twice in
+            // a row
+            if (result.code() == 503 || result.code() == 500) {
+                retryOrFail(() -> new SafeRuntimeException("Retries exhausted"));
+                return;
+            }
+
             delegate.set(result);
         }
 
         @Override
         public void onFailure(Throwable throwable) {
-            if (failures.incrementAndGet() < maxRetries) {
-                Futures.addCallback(runnable.get(), this, DIRECT_EXECUTOR);
+            retryOrFail(() -> throwable);
+        }
+
+        private void retryOrFail(Supplier<Throwable> throwable) {
```

Comment:
I think these failures should always be logged at info or warn level, so we can do away with the supplier